File: scripts/import_hardcover_want_to_read.py
#!/usr/bin/env python3
"""One-off import: pulls your Hardcover "Want to Read" shelf into Media Garden book notes.

Usage:
    HARDCOVER_TOKEN=<token> VAULT_PATH=/path/to/vault python3 scripts/import_hardcover_want_to_read.py

Optional: BASE_FOLDER=Media (must match the plugin's "Base folder" setting, default "Media")
"""

import json
import os
import re
import sys
import urllib.request
from datetime import datetime
from pathlib import Path

ENDPOINT = "https://api.hardcover.app/v1/graphql"


def gql(token: str, query: str, variables: dict) -> dict:
    auth = token if token.startswith("Bearer ") else f"Bearer {token}"
    body = json.dumps({"query": query, "variables": variables}).encode("utf-8")
    req = urllib.request.Request(
        ENDPOINT,
        data=body,
        method="POST",
        headers={
            "Content-Type": "application/json",
            "Authorization": auth,
            "User-Agent": "media-garden-plugin import script (personal use)",
        },
    )
    with urllib.request.urlopen(req) as res:
        payload = json.load(res)
    if payload.get("errors"):
        raise RuntimeError(payload["errors"][0]["message"])
    return payload["data"]


def sanitize(name: str) -> str:
    return re.sub(r'[\\/:*?"<>|]', "", name).strip()


def now_timestamp() -> str:
    return datetime.now().strftime("%Y-%m-%dT%H:%M")


def yaml_string(value) -> str:
    return '"' + str(value).replace("\\", "\\\\").replace('"', '\\"') + '"'


def literary_type(literary_type_id) -> str | None:
    return {1: "Fiction", 2: "Nonfiction"}.get(literary_type_id)


def build_frontmatter(book: dict) -> str:
    author_list = book.get("contributions") or []
    author = author_list[0]["author"]["name"] if author_list and author_list[0].get("author") else None
    cover = (book.get("image") or {}).get("url")
    rating = book.get("rating")
    lit_type = literary_type(book.get("literary_type_id"))

    lines = [
        "ContentType: media",
        "type: book",
        f"title: {yaml_string(book['title'])}",
        f"year: {book.get('release_year') or ''}",
        "genres: []",
        f"cover: {yaml_string(cover) if cover else ''}",
        f"external_rating: {round(rating * 2, 2) if rating else ''}",
        f"created: {now_timestamp()}",
        "status: planning",
        'rating: ""',
        'date_started: ""',
        'date_finished: ""',
        "digital_copy: false",
        f"author: {yaml_string(author) if author else ''}",
        f"pages: {book.get('pages') or ''}",
        f"literary_type: {yaml_string(lit_type) if lit_type else ''}",
        "physical_copy: false",
    ]
    return "\n".join(lines)


def build_body(book: dict) -> str:
    return "\n".join(
        ["## Summary", "", book.get("description") or "", "", "## Quotes", "", "## Notes", "", "## Review", ""]
    )


def main() -> None:
    token = os.environ.get("HARDCOVER_TOKEN")
    vault_path = os.environ.get("VAULT_PATH")
    base_folder = os.environ.get("BASE_FOLDER", "Media")

    if not token or not vault_path:
        print(
            "Usage: HARDCOVER_TOKEN=<token> VAULT_PATH=<path to vault> "
            "python3 scripts/import_hardcover_want_to_read.py",
            file=sys.stderr,
        )
        sys.exit(1)

    me = gql(token, "query { me { id } }", {})
    user_id = me["me"][0]["id"]
    if not user_id:
        raise RuntimeError("Could not resolve your Hardcover user id from the `me` query.")

    data = gql(
        token,
        """
        query WantToRead($userId: Int!) {
            user_books(where: { user_id: { _eq: $userId }, status_id: { _eq: 1 } }, limit: 500) {
                book {
                    title
                    description
                    rating
                    pages
                    release_year
                    image { url }
                    contributions { author { name } }
                    literary_type_id
                }
            }
        }
        """,
        {"userId": user_id},
    )

    books = [ub["book"] for ub in (data.get("user_books") or []) if ub.get("book")]
    print(f'Found {len(books)} "Want to Read" books.')

    folder = Path(vault_path) / base_folder / "Books"
    folder.mkdir(parents=True, exist_ok=True)

    created = skipped = 0
    for book in books:
        title_part = sanitize(book["title"]) or "Untitled"
        year = book.get("release_year")
        filename = f"{title_part} ({year}).md" if year else f"{title_part}.md"
        file_path = folder / filename

        if file_path.exists():
            skipped += 1
            continue

        content = f"---\n{build_frontmatter(book)}\n---\n\n{build_body(book)}"
        file_path.write_text(content, encoding="utf-8")
        created += 1
        print(f"Created: {filename}")

    print(f"Done. {created} created, {skipped} already existed and were left untouched.")


if __name__ == "__main__":
    main()

File: scripts/import-hardcover-want-to-read.mjs
#!/usr/bin/env node
// One-off import: pulls your Hardcover "Want to Read" shelf into Media Garden book notes.
// Usage:
//   HARDCOVER_TOKEN=<token> VAULT_PATH=/path/to/vault node scripts/import-hardcover-want-to-read.mjs
// Optional: BASE_FOLDER=Media (must match the plugin's "Base folder" setting, default "Media")

import { mkdir, writeFile, access } from "node:fs/promises";
import path from "node:path";

async function gql(token, query, variables) {
	const res = await fetch("https://api.hardcover.app/v1/graphql", {
		method: "POST",
		headers: {
			"Content-Type": "application/json",
			Authorization: token.startsWith("Bearer ") ? token : `Bearer ${token}`,
			"User-Agent": "media-garden-plugin import script (personal use)",
		},
		body: JSON.stringify({ query, variables }),
	});
	const json = await res.json();
	if (json.errors?.length) throw new Error(json.errors[0].message);
	return json.data;
}

export function sanitize(name) {
	return name.replace(/[\\/:*?"<>|]/g, "").trim();
}

function nowTimestamp() {
	const d = new Date();
	const pad = (n) => String(n).padStart(2, "0");
	return `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())}T${pad(d.getHours())}:${pad(d.getMinutes())}`;
}

function yamlString(s) {
	return `"${String(s).replace(/\\/g, "\\\\").replace(/"/g, '\\"')}"`;
}

export function buildFrontmatter(book) {
	return [
		`ContentType: media`,
		`type: book`,
		`title: ${yamlString(book.title)}`,
		`year: ${book.release_year ?? ""}`,
		`genres: []`,
		`cover: ${book.image?.url ? yamlString(book.image.url) : ""}`,
		`external_rating: ${book.rating ? (book.rating * 2).toFixed(2) : ""}`,
		`created: ${nowTimestamp()}`,
		`status: planning`,
		`rating: ""`,
		`date_started: ""`,
		`date_finished: ""`,
		`digital_copy: false`,
		`author: ${book.contributions?.[0]?.author?.name ? yamlString(book.contributions[0].author.name) : ""}`,
		`pages: ${book.pages ?? ""}`,
		`physical_copy: false`,
	].join("\n");
}

export function buildBody(book) {
	return ["## Summary", "", book.description ?? "", "", "## Quotes", "", "## Notes", "", "## Review", ""].join("\n");
}

async function fileExists(p) {
	try {
		await access(p);
		return true;
	} catch {
		return false;
	}
}

async function main() {
	const TOKEN = process.env.HARDCOVER_TOKEN;
	const VAULT_PATH = process.env.VAULT_PATH;
	const BASE_FOLDER = process.env.BASE_FOLDER || "Media";

	if (!TOKEN || !VAULT_PATH) {
		console.error(
			"Usage: HARDCOVER_TOKEN=<token> VAULT_PATH=<path to vault> node scripts/import-hardcover-want-to-read.mjs"
		);
		process.exit(1);
	}

	const me = await gql(TOKEN, `query { me { id } }`, {});
	const userId = me.me?.id;
	if (!userId) throw new Error("Could not resolve your Hardcover user id from the `me` query.");

	const data = await gql(
		TOKEN,
		`query WantToRead($userId: Int!) {
			user_books(where: { user_id: { _eq: $userId }, status_id: { _eq: 1 } }, limit: 500) {
				book {
					title
					description
					rating
					pages
					release_year
					image { url }
					contributions { author { name } }
				}
			}
		}`,
		{ userId }
	);

	const books = (data.user_books ?? []).map((ub) => ub.book).filter(Boolean);
	console.log(`Found ${books.length} "Want to Read" books.`);

	const folder = path.join(VAULT_PATH, BASE_FOLDER, "Books");
	await mkdir(folder, { recursive: true });

	let created = 0;
	let skipped = 0;
	for (const book of books) {
		const titlePart = sanitize(book.title) || "Untitled";
		const filename = book.release_year ? `${titlePart} (${book.release_year}).md` : `${titlePart}.md`;
		const filePath = path.join(folder, filename);

		if (await fileExists(filePath)) {
			skipped++;
			continue;
		}

		const content = `---\n${buildFrontmatter(book)}\n---\n\n${buildBody(book)}`;
		await writeFile(filePath, content, "utf8");
		created++;
		console.log(`Created: ${filename}`);
	}

	console.log(`Done. ${created} created, ${skipped} already existed and were left untouched.`);
}

if (import.meta.url === `file://${process.argv[1]}`) {
	main().catch((err) => {
		console.error("Import failed:", err.message);
		process.exit(1);
	});
}
